### PR TITLE
Implement peekJson() for JsonValueReader

### PR DESCRIPTION
**Pull request by swankjesse**
Related to https://github.com/square/moshi/issues/672

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/JsonValueReader.java [hunk 1]
```diff
@@ -49,17 +50,31 @@
...
+    stack = new Object[32];
     stack[stackSize++] = root;
   }
 
+  /** Copy-constructor makes a deep copy for peeking. */
+  JsonValueReader(JsonValueReader copyFrom) {
+    super(copyFrom);
+
+    stack = copyFrom.stack.clone();
+    for (int i = 0; i < stackSize; i++) {
```
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/JsonValueReader.java]: style nit: use enhanced for loop for the array?

File: moshi/src/main/java/com/squareup/moshi/JsonValueReader.java [hunk 2]
```diff
@@ -374,4 +393,32 @@ private void remove() {
...
+    @Override public Object next() {
+      return list[next++];
+    }
+
+    @Override public void remove() {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override protected JsonIterator clone() {
+      return new JsonIterator(endToken, list, next);
```
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/JsonValueReader.java]: worth a comment to state that the `list` doesn't need a clone?

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java [hunk 3]
```diff
@@ -985,4 +985,98 @@ JsonReader newReader(String json) throws IOException {
...
+    reader.endObject();
+    assertThat(reader.peek()).isEqualTo(JsonReader.Token.END_DOCUMENT);
+  }
+
+  @Test public void peekJsonReader() throws IOException {
+    JsonReader reader = newReader("[12,34,{\"a\":56,\"b\":78},90]");
+    assumeTrue(reader instanceof JsonValueReader); // Not implemented for JsonUtf8Reader yet!
+    for (int i = 0; i < 12; i++) {
+      readPeek12Steps(reader.peekJson(), i, 12);
+      readPeek12Steps(reader, i, i + 1);
```
Comment (NightlyNexus) [inline moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: this _looks_ like the from and until indices are unused?
i don't follow. what's this test case behavior?

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java [hunk 4]
```diff
@@ -985,4 +985,98 @@ JsonReader newReader(String json) throws IOException {
...
+    }
+  }
+
+  /**
+   * Read a fragment of {@code reader}. This assumes the fixed document defined in {@link
+   * #peekJsonReader} and reads a range of it on each call.
+   */
+  private void readPeek12Steps(JsonReader reader, int from, int until) throws IOException {
```
Comment (NightlyNexus) [inline moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: is this simply `if (from == until) return;`?

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java [hunk 5]
```diff
@@ -461,6 +470,32 @@ public final void setFailOnUnknown(boolean failOnUnknown) {
     }
   }
 
+  /**
+   * Returns a new {@code JsonReader} that can read data from this {@code JsonReader} without
```
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: nice! will we need a note to say that behavior is undefined if the original gets ahead of the peeked reader?

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java [hunk 6]
```diff
@@ -200,6 +200,15 @@
     // Package-private to control subclasses.
```
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: nit: should we move the other assignments into this constructor like JsonValueReader's `stack` did?

File: moshi/src/main/java/com/squareup/moshi/JsonValueReader.java (hunk 1 above)
Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonValueReader.java]: it’s slightly different semantics ’cause the loop doesn’t span the physical storage!

File: moshi/src/main/java/com/squareup/moshi/JsonValueReader.java (hunk 2 above)
Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonValueReader.java]: Renamed to `array`!

Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonValueReader.java]: Fixed.

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java (hunk 3 above)
Comment (swankjesse) [reply moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: The real reader uses 12 operations to read a JSON document. Before each of the real reader’s operations we create a peeking reader and let it read the rest of the document.

Each `.` in here is a single read operation.
 
```
PEEK ............
REAL .           
PEEK  ...........
REAL  .          
PEEK   ..........
REAL   .         
PEEK    .........
REAL    .        
PEEK     ........
REAL     .       
PEEK      .......
REAL      .      
PEEK       ......
REAL       .     
PEEK        .....
REAL        .    
PEEK         ....
REAL         .   
PEEK          ...
REAL          .  
PEEK           ..
REAL           . 
PEEK            .
REAL            .
```

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java (hunk 4 above)
Comment (swankjesse) [reply moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: Not exactly.

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java (hunk 3 above)
Comment (NightlyNexus) [reply moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: that's what i thought it was supposed to be, and i read this at least two dozen times looking for a loop until i finally realized i was glazing over the fall-throughs. switch statements get me every time!

fantastic! 👍

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java (hunk 4 above)
Comment (NightlyNexus) [reply moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: forgot switch statements today. 👍

File: moshi/src/main/java/com/squareup/moshi/JsonValueReader.java (hunk 1 above)
Comment (NightlyNexus) [reply moshi/src/main/java/com/squareup/moshi/JsonValueReader.java]: whoops, yes, ofc.

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java (hunk 5 above)
Comment (NightlyNexus) [reply moshi/src/main/java/com/squareup/moshi/JsonReader.java]: yeah, mirroring with Okio is perfect!

**Pull request merged**